pipeline/DESIGN.md

```diff
@@ -0,0 +1,190 @@
+| dimensions  | map(string,bytes) | Set of key values dimension pairs of the stream                                                 |
+| event_cid   | bytes             | Cid of the event                                                                                |
+| event_type  | u8                | One of Data (0) or Time (1)                                                                     |
+| data        | bytes             | DAG-JSON encoding of the event payload                                                          |
+| previous    | list(bytes)       | List of CID previous to this event                                                              |
+
+
+### Transformation
+
+Raw events are transformed into a flattened structure where stream dimensions and conclusions about data are added to the data.
+
+Conclusions include:
+
+* The event has a valid signature
+* The dimensions and controller of the event
+
+This table joins the raw_events, chain_proofs, and streams tables in order to make the conclusions about the raw events.
+
+## model_schemas
+
+The model_schemas table contains a row for each model and contains the complete resolved schema of the model.
+
+TBD how this table is populated and its schema.
+This table may be able to be combined with the streams table. Should we?
+
+## event_states
+
+The event_states table contains a row for each event in a stream and the state of the document at that point in the stream.
+
+### Features
+
+* Access to the full history of states for a stream
+* Validation of the model schema
+
+### Schema
+
+| Column      | Type              | Description                                                                                     |
+| ------      | ----              | -----------                                                                                     |
+| index       | u64               | Order of this event. Index is always greater than the index of any previous event in the stream |
+| stream_cid  | bytes             | Cid of the stream                                                                               |
+| stream_type | u8                | Stream type                                                                                     |
+| controller  | string            | Controller of the stream                                                                        |
+| dimensions  | map(string,bytes) | Set of key values dimension pairs of the stream                                                 |
+| event_cid   | bytes             | Cid of the event                                                                                |
+| event_type  | u8                | One of Data (0) or Time (1)                                                                     |
+| state       | string            | JSON encoding of the event state                                                                |
```

Comment:
Is this always going to be JSON? 

I guess I'm wondering if this type should be bytes that are interpreted based on stream type? Either json or whatever we want it to be? We could always multibase encode arrow or avro or whatever into a string too so bytes vs string might not matter (and it's possible I'm just missing something xd)

Reply:
I have gone back on forth on this, making it a string makes it much easier to work with as all the tooling will just display it. However to make it more generic it should be bytes. Let me see if the built in json functions work with bytes columns and then if so I'll update this to be bytes.

Reply:
When with bytes. It simple to cast to a string when needed i.e. `select state::varchar->>'data' from event_states`